File: LeetCode/HashMap/python/wordPattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split(" ")
        
        if len(words) != len(pattern):
            return False

        charToWord = {}
        wordToChar = {}

        for i in range(len(pattern)):
            c = pattern[i]
            w = words[i]

            if c in charToWord and charToWord[c] != w:
                return False
            
            if w in wordToChar and wordToChar[w] != c:
                return False
            
            charToWord[c] = w
            wordToChar[w] = c

        return True

    @staticmethod
    def testSolution(input) -> None:
        print("Input: pattern: " + input["pattern"])
        print("\ts: " + input["s"])
        print("Expected: " + str(input["expected"]))
        res = Solution().wordPattern(input["pattern"], input["s"])
        print("Result: " + str(res))
        print("PASS" if res == input["expected"] else "FAIL")
        print("-" * 50)

if __name__ == "__main__":
    input = {"pattern": "abba", "s": "dog cat cat dog", "expected": True}
    Solution.testSolution(input)

    input = {"pattern": "abba", "s": "dog cat cat fish", "expected": False}
    Solution.testSolution(input)

    input = {"pattern": "aaaa", "s": "dog cat cat dog", "expected": False}
    Solution.testSolution(input)

File: LeetCode/HashMap/java/wordPattern.java
package LeetCode.HashMap.java;

import java.util.HashMap;
import java.util.Map;

record wordPatternRecord(String pattern, String s, boolean expected) {

}

public class wordPattern {
    public boolean wordPatternSolution(String pattern, String s) {
        // We need to create an array of words from the input s
        String[] words = s.split("\\s+");

        // We need to make sure that the length of the list of words is equal to the
        // length of the number of chars in the pattern string.
        if (words.length != pattern.length()) {
            return false;
        }

        // We'll use a JS Object to keep track of k/v pairs
        Map<Character, String> charToWord = new HashMap<>();
        Map<String, Character> wordToChar = new HashMap<>();

        // Since we know that both the pattern and words array should have the same
        // length, we'll use the pattern lenght to bound the for-loop
        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);
            String w = words[i];

            if (charToWord.containsKey(c) && !charToWord.get(c).equals(w)) {
                return false;
            }

            if (wordToChar.containsKey(w) && !wordToChar.get(w).equals(c)) {
                return false;
            }

            charToWord.put(c, w);
            wordToChar.put(w, c);
        }

        return true;
    }

    public static void main(String[] args) {
        wordPatternRecord input = new wordPatternRecord("abba", "dog cat cat dog", true);
        wordPattern.testSolution(input);

        input = new wordPatternRecord("abba", "dog cat cat fish", false);
        wordPattern.testSolution(input);

        input = new wordPatternRecord("aaaa", "dog cat cat dog", false);
        wordPattern.testSolution(input);
    }

    private static void testSolution(wordPatternRecord input) {
        System.out.println("Input: pattern: " + input.pattern());
        System.out.println("\ts: " + input.s());
        System.out.println("Expected: " + input.expected());
        boolean res = new wordPattern().wordPatternSolution(input.pattern(), input.s());
        System.out.println("Result: " + res);
        System.out.println(res == input.expected() ? "PASS" : "FAIL");
        System.out.println("-".repeat(50));
    }
}
